calling-apis: content (#15)
Ada-Developers-Academy/core-unit-3: calling-apis/get-requests-with-axios.md

# GET Requests With axios

## Goals

- Request data from an API to render client-side using JavaScript
- Render data from server without reloading a page
- Use axios to make HTTP GET requests

## Introduction

Jamel is working on the JavaScript front-end of his app called DogSpotter, where he maps dogs he spots to the neighborhoods he lives in.

Jamel needs to get a source of dog breeds. He's not an expert on dogs, but he knows that there's an API that will help him get this data: [The Dog CEO API](https://dog.ceo/dog-api/documentation/).

He's made API calls in other programming languages before, but he needs to get up to speed on how to call this one now. What can Jamel do to learn how to make API calls?

After doing some research, Jamel learns two things:

- He will choose to use a Node package named `axios` to make API calls. He needs to learn the syntax for using this library.
- API calls in JavaScript typically execute _asynchronously_. Jamel will learn a little bit about what that means to help him understand his code.

> <!-- To make a `GET` request, our goal is to answer the following questions:
> 
> 1. What is the library/tool that will help us make `GET` requests in JavaScript?
> 1. How do we make the `GET` request, which is now an asynchronous call?
> 1. What will happen if the `GET` request finishes successfully, and gives us back a successful response?
> 1. What will happen if the `GET` request does **not** finish successfully, and gives us back a failure response code? -->

Author: Woops, yeah, this was just part of textbook-curriculum. I used it as a reference to write some of this material, this is a stray comment!

```suggestion
```

<!-- prettier-ignore-start -->
### !challenge
* type: multiple-choice
* id: 7cbb2b75
* title: GET Requests With axios
##### !question

In the case of handling an unsuccessful request, which of the following options gives back the HTTP response body, which contains the data that the API sent back?

##### !end-question
##### !options

* `error`
* `error.response.status`
* `error.response.data`

##### !end-options
##### !answer

* `error.response.data`

##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->

### `finally`

`axios` supports a `finally` clause, which will run after either the `then` or `catch` no matter what.

The example syntax looks like this:

```js
const axios = require("axios");

axios
  .get("https://dog.ceo/api/breeds/list/all")
  .then((response) => {
    console.log("success!");
  })
  .catch((error) => {
    console.log("error!");
  })
  .finally(() => {
    console.log("this is always executed, no matter what!");
  });
```

## Query Params

As part of Jamel's DogSpotter app, he needs to find the latitude and longitude of Seattle, Washington, USA. He'll use the [LocationIQ Geocoding API](https://locationiq.com/sandbox/geocoding/forward).

> After looking at [the docs for Search/Forward Geocoding](https://locationiq.com/docs), Jamel determined that his request should look like this:

Reviewer: keep present voicing
```suggestion
After looking at [the docs for Search/Forward Geocoding](https://locationiq.com/docs), Jamel determines that his request should look like this:
```

- `GET` request to `https://us1.locationiq.com/v1/search.php`
- Query Params:
  - `key`: the API key

>   - `q`: A search term. In this case, `"Seattle, Washington, USA"`
>   - `format`: `"json"`

Reviewer: updates to preferred quotes
```suggestion
  - `q`: A search term. In this case, `'Seattle, Washington, USA'`
  - `format`: `'json'`
```

He expects a response body that looks like:

```json
[

>   {
>     "place_id": "332439918643",
>     "licence": "https://locationiq.com/attribution",
>     "lat": "47.673455",
>     "lon": "-122.331537",
>     "display_name": "Seattle, King County, Washington, USA",
>     "boundingbox": ["47.495514", "47.734165", "-122.435956", "-122.236044"],
>     "importance": 0.25
>   }

Reviewer: looks like this might be an old request? not sure what caused some of the lat/lon changes (my guess is that maybe they're using a different reference ellipsoid???)

Feel free to cut down some of the fields in this response if they are more noisy than helpful.

Also, I noticed that even when making the exact same request repeated times, I would sometimes gets 1 result, and other times several (some stuff in Kent and elsewhere).

```suggestion
  {
    "place_id": "235549103",
    "licence": "https://locationiq.com/attribution",
    "osm_type": "relation",
    "osm_id": "237385",
    "boundingbox": [ "47.4810022", "47.7341357", "-122.459696", "-122.224433" ],
    "lat": "47.6038321",
    "lon": "-122.3300624",
    "display_name": "Seattle, King County, Washington, USA",
    "class": "place",
    "type": "city",
    "importance": 0.8929791735643791,
    "icon": "https://locationiq.org/static/images/mapicons/poi_place_city.p.20.png"
  },

  // possibly other matches
```

]
```

> To add query params to our `GET` request, Jamel can research the `axios` documentation.

Reviewer: search the documentation doesn't add the params, but _does_ tell us _how_ to add the params.
```suggestion
To figure out how to add query params to our `GET` request, Jamel can research the `axios` documentation.
```

He learns that to add query params, he utilizes `axios.get()`'s second parameter. This function takes in an optional "config" object.

Within this "config" object, there we can add in a key `params`. The value of `params` should be another object.

This `params` object should contain the key-value pairs we want to send as query params.

Jamel's code eventually looks [like this](https://replit.com/@adacore/GET-Requests-With-axios-Demo#index.js):

```js

> const axios = require("axios");
> 
> axios
>   .get("https://us1.locationiq.com/v1/search.php", {
>     params: {
>       key: "... jamel's api key ...",
>       q: "Seattle, Washington, USA",
>       format: "json",
>     },
>   })
>   .then((response) => {
>     console.log("success!", response.data);
>   })
>   .catch((error) => {
>     console.log("error!", error.response.data);
>   });

Reviewer: updates to preferred quotes
also, I really think we should show an example that doesn't hard code the api key, so I updated the key line here as well. following comments introduce a new section that discusses how to set up the key in replit.

this needs to be fixed in the replit as well

```suggestion
const axios = require('axios');

axios
  .get('https://us1.locationiq.com/v1/search.php', {
    params: {
      key: process.env['api_key'],  // discussed below
      q: 'Seattle, Washington, USA',
      format: 'json',
    },
  })
  .then((response) => {
    console.log('success!', response.data);
  })
  .catch((error) => {
    console.log('error!', error.response.data);
  });
```

```

> 

Reviewer: Adds a section detailing how to set up the api key secret in replit.
```suggestion

Before trying to run this code ourselves, we should examine how Jamel configured his LocationIQ API key.

Recall that we should always try to keep our API keys private. During local development, we have created `.env` files to hold these values to help avoid accidentally committing them to Git. Other environments have provided management consoles where we can set secrets so that they are kept safe from prying eyes!

Replit.com provides a UI for creating secrets, which we can then access through the `process.env` object.

To set up our secret in our REPL, we click the Secrets icon (shaped like a padlock).

![The Replit.com interface showing the padlock-shaped icon highlighted. The hover text describes the icon as "Secrets (Environment variables)."](./../assets/calling-apis_get-requests-with-axios_secrets.png)  
_Fig. Locating the Secrets panel in Replit.com_

Then we fill in the key name with an appropriate name (we used `api_key`), paste our LocationIQ API key as the value, then click "Add new secret."

![The Replit.com interface showing the Secrets panel. The key field has "api_key" entered, and a placeholder in the value field of "YOUR_LOCATIONIQ_API_KEY" is displayed.](./../assets/calling-apis_get-requests-with-axios_secrets-api-key.png)  
_Fig. Filling in our secret API key_

Now our API key will be accessible in our code through `process.env['api_key']`!

```

> When he runs it, he indeed gets back the successful response he hoped for!

Reviewer: updates to transition from the new section
```suggestion
Following these steps, Jamel securely set up his API key. Now when he runs his code, he indeed gets back the successful response he hoped for!
```

## Observing Asynchronicity

> This API call is made _asynchronously_. This asynchronous nature will affect Jamel approaches his project. But first, how can he observe the asynchronous property of this code?
> 
> Let's run this code and observe:

Reviewer: "asynchronous property" sounds like we might be talking about a field on an object. That's where I'd want to say "asynchronous nature", but that sounds redundant with the first part, so that became "asynchronous behavior"

call out what the expectation (which we will subvert) is

```suggestion
This API call is made _asynchronously_. This asynchronous behavior will affect how Jamel approaches his project. But first, how can he observe the asynchronous nature of this code?

Let's run this code, which appears to make an API call and then prints some lyrics:
```

```js

> const axios = require("axios");
> 
> axios.get("https://dog.ceo/api/breeds/list/all")
>   .then(() => {
>     console.log('success!');
>   })
>   .catch(() => {
>     console.log('error!');
>   });
> 
> console.log("I broke free on a Saturday morning");
> console.log("I put the pedal to the floor");
> console.log("Headed north on Mills Avenue");
> console.log("And listened to the engine roar");

Reviewer: updates to preferred quotes

this needs to be fixed in the replit as well

```suggestion
const axios = require('axios');

axios.get('https://dog.ceo/api/breeds/list/all')
  .then(() => {
    console.log('success!');
  })
  .catch(() => {
    console.log('error!');
  });

console.log('I broke free on a Saturday morning');
console.log('I put the pedal to the floor');
console.log('Headed north on Mills Avenue');
console.log('And listened to the engine roar');
```

```

When we run this code, we should see this output:

```
I broke free on a Saturday morning
I put the pedal to the floor
Headed north on Mills Avenue
And listened to the engine roar
success!
```

Our console prints out the lyrics at the bottom first, **before** our API call prints "success!" But our song lyrics were written _below_ the lines where we made our API call.

Our JavaScript still gets executed from top to bottom. However, when we make our GET request, our program **does not wait for the HTTP response to come back**. Instead, after sending the `GET` request, the program **proceeds** and continues to execute the next lines of code.

### Working With Asynchronous Code

Because this is asynchronous, we sometimes need to tailor our code to anticipate that our response may come back _after_ the rest of our program executes.

Observe this bug:

```js

> const axios = require("axios");
> 
> const dogGreeting = "bark";
> let dogImageURL;
> 
> axios
>   .get("https://dog.ceo/api/breeds/image/random")
>   .then((response) => {
>     dogImageURL = response.data.message;
>   })
>   .catch((error) => {
>     dogImageURL = "There was an error";
>   });
> 
> console.log(`My favorite dog greeting is ${dogGreeting}`);
> console.log(`A random dog image URL is ${dogImageURL}`);

Reviewer: updates to preferred quotes

consider adding this to the replit

```suggestion
const axios = require('axios');

const dogGreeting = 'bark';
let dogImageURL;

axios
  .get('https://dog.ceo/api/breeds/image/random')
  .then((response) => {
    dogImageURL = response.data.message;
  })
  .catch((error) => {
    dogImageURL = 'There was an error';
  });

console.log(`My favorite dog greeting is ${dogGreeting}`);
console.log(`A random dog image URL is ${dogImageURL}`);
```

```

> Reading through this code from top-to-bottom, without understanding asynchronosity, would lead us to believe that:
> 
> 1. We declare `let dogImageURL` at the top
> 1. **Then,** we make an API call, and then assign the value of the API's response data to `dogImageURL`
> 1. **After that**, we print to the terminal with `console.log` the value of `dogImageURL`
> 
> However, because the API call happens asynchronously, our program doesn't wait for the API response to come back before proceeding with the program. The result is that our `dogImageURL` comes back as `undefined`.

Reviewer: asynchronosity typo
reorder the sentence describing the print statement (keep the direct object closer to the verb)
at the point when the print occurs, the result hasn't yet "come back", so restate to indicate that it hasn't been changed yet

```suggestion
Reading through this code from top-to-bottom, without understanding asynchronicity, would lead us to believe that:

1. We declare `let dogImageURL` at the top
1. **Then,** we make an API call, and assign the value of the API's response data to `dogImageURL`
1. **After that**, we print the value of `dogImageURL` to the terminal using `console.log()`

However, because the API call happens asynchronously, our program doesn't wait for the API response to come back before proceeding with the program. The result is that when we try to output our `dogImageURL` it is still `undefined`!
```

When we need to solve problems with API calls, we should anticipate these scenarios, and consider what logic needs to be moved into our `then` and `catch` blocks.

### !callout-info

## Give Me More About Handling Asynchronous Code!

> JavaScript features a lot of ways to work with asynchronous code. Different search terms to use to research more could be "JavaScript Promises" and "JavaScript async await."

Reviewer: feels like this could benefit from a FYC
```suggestion
JavaScript features a lot of ways to work with asynchronous code. Different search terms to use to research more could be "JavaScript Promises" and "JavaScript async await." Follow your curiosity!
```
